core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
+        if(hasSharePermission) DBIO.successful(hasSharePermission)
+        else rawlsGroupQuery.listGroupsForUser(RawlsUserRef(subjectId)).flatMap { userGroups =>
+          val groupNames = userGroups.map(_.groupName.value)
+          workspaceGroupShareQuery.filter(rec => (rec.groupName) inSet(groupNames)).countDistinct.result.map(rows => rows > 0)
+        }
+      }
+    }
+
+    def listEmailsAndAccessLevel(workspaceContext: SlickWorkspaceContext): ReadAction[Seq[(String, WorkspaceAccessLevel, Boolean)]] = {
       val accessAndUserEmail = (for {
         access <- workspaceAccessQuery if (access.workspaceId === workspaceContext.workspaceId && access.isRealmAcl === false)
         group <- rawlsGroupQuery if (access.groupName === group.groupName)
         userGroup <- groupUsersQuery if (group.groupName === userGroup.groupName)
         user <- rawlsUserQuery if (user.userSubjectId === userGroup.userSubjectId)
-      } yield (access, user)).map { case (access, user) => (access.accessLevel, user.userEmail) }
+      } yield (access, user)).map { case (access, user) => (access.accessLevel, user.userEmail, user.userSubjectId) }
 
       val accessAndSubGroupEmail = (for {
         access <- workspaceAccessQuery if (access.workspaceId === workspaceContext.workspaceId && access.isRealmAcl === false)
         group <- rawlsGroupQuery if (access.groupName === group.groupName)
         subGroupGroup <- groupSubgroupsQuery if (group.groupName === subGroupGroup.parentGroupName)
         subGroup <- rawlsGroupQuery if (subGroup.groupName === subGroupGroup.childGroupName)
-      } yield (access, subGroup)).map { case (access, subGroup) => (access.accessLevel, subGroup.groupEmail) }
+      } yield (access, subGroup)).map { case (access, subGroup) => (access.accessLevel, subGroup.groupEmail, subGroup.groupName) }
 
-      (accessAndUserEmail union accessAndSubGroupEmail).result.map(_.map { case (access, email) => 
-        (email, WorkspaceAccessLevels.withName(access))
-      })
+      /*  The left join here is important. Since we are only going to store share-permission records for users and groups that have been
+          explicitly granted that ability, we want to be able to return all users and groups even if they don't have share permissions.
+          After the join, a null column equates to false, and a non-null column equates to true. This conversion is done with permission.isDefined
+       */
+      val userShareQuery = accessAndUserEmail.joinLeft(workspaceUserShareQuery).on(_._3 === _.userSubjectId).map { case ((accessLevel, userEmail, _), shareRecord) => (accessLevel, userEmail, shareRecord.isDefined) }
+      val groupShareQuery = accessAndSubGroupEmail.joinLeft(workspaceGroupShareQuery).on(_._3 === _.groupName).map { case ((accessLevel, groupEmail, _), shareRecord) => (accessLevel, groupEmail, shareRecord.isDefined) }
```

Comment:
The SQL query that slick actually generates:

```
SELECT 
    x2.`access_level`, x3.`EMAIL`, x4.`workspace_id` IS NOT NULL
FROM
    `WORKSPACE_ACCESS` x2
        INNER JOIN
    (`GROUP` x5
    INNER JOIN (`GROUP_SUBGROUPS` x6
    INNER JOIN `GROUP` x3 ON x3.`NAME` = x6.`CHILD_NAME`) ON x5.`NAME` = x6.`PARENT_NAME`) ON ((x2.`workspace_id` = 0xE5D5CA221F0146DF9F92551963A9A1)
        AND (x2.`is_realm_acl` = FALSE))
        AND (x2.`group_name` = x5.`NAME`)
        LEFT OUTER JOIN
    `WORKSPACE_GROUP_SHARE` x4 ON x3.`NAME` = x4.`group_name`;
```

explain plan:

<img width="783" alt="screen shot 2017-01-25 at 9 32 22 am" src="https://cloud.githubusercontent.com/assets/7257391/22294406/4137a8aa-e2e1-11e6-87ea-de0662de7b8d.png">